### ISSUE TITLE
Python Bindings

### ISSUE THREAD

**Issue by rsyring** [couchbase/couchbase-lite-core#91]
Just wondering if you would consider Python bindings for this library.  

I'm currently planning a project that essentially needs a local couchdb server to save data too so that it can be replicated out to other stores that implement the couchdb sync protocol.  But really, a full CouchDB install seems overkill.  What I really want is more akin to an SQLite db.  If I understand it correctly, that is exactly what couchbase-lite is.  Since the project will be primarily in Python, I figured I'd get your thoughts on Python bindings.

Thanks.

**Comment by snej** [comment]
We've talked about a Python binding for the purpose of writing a lot of functional tests, since they'd be quicker/easier to write than in C/C++. But there's no actual plan to do one. If you wanted to tackle it, that would be great.

> so that it can be replicated out to other stores that implement the couchdb sync protocol.

We're not implementing the CouchDB protocol in LiteCore. We have a new protocol based on WebSockets that's a lot more efficient. If you need the CouchDB protocol, you should take a look at Couchbase Lite 1.x instead.

**Comment by rsyring** [comment]
Thanks for the quick reply.

> for the purpose of writing a lot of functional tests

Do you have any functional tests currently written you could point me at that would be good/simple candidates for doing in Python instead?

> We have a new protocol based on WebSockets that's a lot more efficient

What will that be compatible with?  That is, what other servers/libraries will speak that protocol?

**Comment by snej** [comment]
@sethrosetter, our QA lead, would know best about functional tests. We don't have any yet for LiteCore or Couchbase Lite 2.0, but we do for 1.x.

The new replication protocol will work with Couchbase Sync Gateway 2.0. (It's already implemented in the latest developer build.) Anyone else is welcome to implement it; it's an [open, documented protocol](https://github.com/couchbase/couchbase-lite-core/wiki/Replication-Protocol).

**Comment by sethrosetter** [comment]
@rsyring Python bindings would be awesome. At the moment, we are planning to do most of testing for 2.0 at the platform level (i.e. Swift, Java, .NET), so it is not a high priority at the moment. I know @borrrden has created a .NET binding to LiteCore and may have some advice or be able to share some code with you for reference. 

This is our QE test framework - https://github.com/couchbaselabs/mobile-testkit

**Comment by borrrden** [comment]
FYI The .NET bindings are part of this repo, as well as scripts (which happen to be [horrendously ugly thanks to me writing them] python) to generate them.  The script acts in two stages.  The [first stage](https://github.com/couchbase/couchbase-lite-core/blob/master/C/include/parse_API.py) can be used as is to generate an abstract representation of the bindings surface area.  

The [second stage](https://github.com/couchbase/couchbase-lite-core/blob/master/CSharp/src/LiteCore.Shared/Interop/gen_bindings.py) is specific to C#, and this would need to be adapted for python to do whatever is needed for python to call into a C function (variable marshalling, etc).  Alternatively, the generated files are all the the same directory as the second stage script (with the suffix `_native` indicating that they are autogenerated binding methods.  The `_defs` files are directly generated via another script that is specific to C#).

To summarize, I am not sure what python needs to do in this case but the methods that need binding are all grouped together in one folder (the C/include/ directory)

**Comment by stcleezy** [comment]
We are also extremely interested in using python bindings for LiteCore in the embedded world. We have built a very basic POC using ctypes with some success. Is there any effort, either within Couchbase or in the open source community to fully support this?

**Comment by borrrden** [comment]
There is no current effort within Couchbase, and as far as I know no efforts by the open source community to add python bindings, sorry!  Maybe you could be the one to start the momentum for it in the open source community, but take a look at the [contributors page](https://github.com/couchbase/couchbase-lite-core/graphs/contributors) and consider that information.

**Comment by stcleezy** [comment]
Thanks for the reply. What about javascript, not specific to React Native? Any official plans for bindings?

**Comment by snej** [comment]
I'd say a JS binding is likely to happen, because right now we have no upgrade story for the people using CBL 1.4 with PhoneGap. But it's not officially on the roadmap yet (nor do I have control over that roadmap!)

**Comment by stcleezy** [comment]
It looks like this is being developed here? https://github.com/couchbaselabs/couchbase-lite-C

**Comment by snej** [comment]
Yup.
